src/System.Data.Common/src/System/Data/Common/DbCommand.cs

```diff
@@ -0,0 +1,265 @@
+            set;
+        }
+
+        abstract protected DbParameterCollection DbParameterCollection
+        { // V1.2.3300
+            get;
+        }
+
+        abstract protected DbTransaction DbTransaction
+        { // V1.2.3300
+            get;
+            set;
+        }
+
+        public abstract bool DesignTimeVisible
+        {
+            get;
+            set;
+        }
+
+        public DbParameterCollection Parameters
+        {
+            get
+            {
+                return DbParameterCollection;
+            }
+        }
+
+        public DbTransaction Transaction
+        {
+            get
+            {
+                return DbTransaction;
+            }
+            set
+            {
+                DbTransaction = value;
+            }
+        }
+
+        abstract public UpdateRowSource UpdatedRowSource
+        {
+            get;
+            set;
+        }
+
+        internal void CancelIgnoreFailure()
+        {
+            // This method is used to route CancellationTokens to the Cancel method.
+            // Cancellation is a suggestion, and exceptions should be ignored
+            // rather than allowed to be unhandled, as there is no way to route
+            // them to the caller.  It would be expected that the error will be
+            // observed anyway from the regular method.  An example is cancelling
+            // an operation on a closed connection.
+            try
+            {
+                Cancel();
+            }
+            catch (Exception)
+            {
+            }
+        }
+
+        abstract public void Cancel();
+
+        public DbParameter CreateParameter()
+        { // V1.2.3300
+            return CreateDbParameter();
+        }
+
+
+        abstract protected DbParameter CreateDbParameter();
+
+        abstract protected DbDataReader ExecuteDbDataReader(CommandBehavior behavior);
+
+        abstract public int ExecuteNonQuery();
+
+        public DbDataReader ExecuteReader()
+        {
+            return (DbDataReader)ExecuteDbDataReader(CommandBehavior.Default);
+        }
+
+
+        public DbDataReader ExecuteReader(CommandBehavior behavior)
+        {
+            return (DbDataReader)ExecuteDbDataReader(behavior);
+        }
+
+
+        public Task<int> ExecuteNonQueryAsync()
+        {
+            return ExecuteNonQueryAsync(CancellationToken.None);
+        }
+
+        public virtual Task<int> ExecuteNonQueryAsync(CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return ADP.CreatedTaskWithCancellation<int>();
+            }
+            else
+            {
+                CancellationTokenRegistration registration = new CancellationTokenRegistration();
+                if (cancellationToken.CanBeCanceled)
+                {
+                    registration = cancellationToken.Register(CancelIgnoreFailure);
+                }
+
+                try
```

Comment:
What happens if cancellation is requested prior to invoking ExecuteNonQuery?  I'm wondering if the implementation appropriately handles that race condition (which could happen because at this point a cancellation request has been wired up to CancelIgnoreFailure).